Sources/UID2/UID2Manager.swift

```diff
@@ -187,13 +232,84 @@ public final actor UID2Manager {
         }
         
         self.identity = validIdentity
-        KeychainManager.shared.saveIdentityToKeychain(validIdentity)
+        KeychainManager.shared.saveIdentityToKeychain(validatedIdentityPackage)
+        
+        await checkIdentityRefresh()
+        await checkIdentityExpiration()
         
         return validIdentity
     }
 
     // MARK: - Refresh and Timer
     
+    private func checkIdentityRefresh() async {
+        refreshJob?.cancel()
+        refreshJob = nil
+
+        if !automaticRefreshEnabled {
+            return
+        }
+        
+        if let identity = identity {
+            // If the identity is already suitable for a refresh, we can do so immediately. Otherwise, we will work out
+            // how long it is until a refresh is required and schedule it accordingly.
+            if await hasExpired(expiry: identity.refreshFrom) {
+                refreshJob = Task {
+                    await refreshToken(identity: identity)
+                }
+            } else {
+                refreshJob = Task {
+                    let now = Date().millisecondsSince1970
+                    let refreshFrom = identity.refreshFrom
+                    let delayInMilliseconds = refreshFrom - now
+                    let delayInNanoseconds = UInt64(delayInMilliseconds * 1000000)
+                    try await Task.sleep(nanoseconds: delayInNanoseconds)
+                    await refreshToken(identity: identity)
+                }
+            }
+        }
+    }
+    
+    /// The identity status can change as we reach specific time events. We want to observe these and make sure that when
+    /// they are reached, we can report them accordingly to our consumer.
+    private func checkIdentityExpiration() async {
+        checkRefreshExpiresJob?.cancel()
+        checkRefreshExpiresJob = nil
+        
+        checkIdentityExpiresJob?.cancel()
+        checkIdentityExpiresJob = nil
+        
+        if !checkExpiration {
+            return
+        }
+        
+        if let identity = identity {
+            
+            // If the expiration time of being able to refresh is in the future, we will schedule a job to detect if we
+            // pass it. This will allow us to reevaluate our state and update accordingly.
+            if await !hasExpired(expiry: identity.refreshExpires) {
+                checkRefreshExpiresJob = Task {
```

Comment:
Just for my own Swift education:
 - Does the `await` on the previous line simply "wait" for the return of `hasExpired`?
 - Does the creation of a new `Task` mean it's executed immediately (async)
 - Can `async` functions only call `async` functions? I'm curious why `hasExpired` needs to be asynchronous given it's just a Long comparison

Reply:
`async/await` is the Swift version of Kotlin `suspend` to produce asynchronous functions.  It enables the system to automatically move work to background threads / queues.  The `await` keyword is there to inform the compiler as well the developer as it make the code easier to read / understand.

`Task` is the bridge between the synchronous and asynchronous worlds so it can start immediately.  It's the Swift version of Kotlin's `Job`.

> Can async functions only call async functions?

Yep.  This is how state mutation is managed.

If you're curious Apple introduced this in 2021 at WWDC.  The video / transcript below is their introduction to how this all works.

https://developer.apple.com/videos/play/wwdc2021/10132/